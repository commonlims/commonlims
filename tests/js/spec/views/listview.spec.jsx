import React from 'react';
import ListView from 'app/components/listView.jsx';
import {shallow} from 'enzyme';
import {Set} from 'immutable';

describe('instantiate listview', function () {
  const oneSample = {
    dataById: {
      1: {
        entity: {
          global_id: 1,
          name: 'mysample',
          location: {
            container: {
              name: 'mycontainer',
            },
          },
          properties: {
            sample_type: {
              value: 'Zombie brain',
            },
          },
        },
        isGroupHeader: false,
      },
    },
    visibleIds: [1],
  };

  const oneSampleType = {
    dataById: {
      1: {
        entity: {
          global_id: 1,
          name: 'Zombie brain',
        },
        isGroupHeader: true,
        children: {
          isFetched: false,
          isExpanded: false,
          cachedIds: [],
        },
      },
    },
    visibleIds: [1],
  };

  const headers = [
    {
      Header: 'Sample name',
      id: 'name',
      accessor: (d) => d.name,
    },
    {
      Header: 'Container',
      id: 'container',
      accessor: (d, isGroupHeader) =>
        isGroupHeader ? null : d.location ? d.location.container.name : '<No location>',
    },
    {
      Header: 'Sample Type',
      id: 'sample_type',
      accessor: (d, isGroupHeader) =>
        isGroupHeader
          ? null
          : d.properties && d.properties.sample_type
          ? d.properties.sample_type.value
          : null,
    },
  ];

  function renderListView(args) {
    const defaultProps = {
      columns: headers,
      errorMessage: '',
      loading: false,
      canSelect: false,
      allVisibleSelected: false,
      toggleAll: jest.fn(),
      toggleSingle: jest.fn(),
      visibleIds: [],
      selectedIds: new Set(),
      dataById: {},
      listActionBar: {},
    };
    const props = {...defaultProps, ...args};
    return shallow(<ListView {...props} />);
  }
  it('listView according to snapshot', () => {
    const wrapper = renderListView({...oneSampleType});
    expect(wrapper.instance()).toMatchSnapshot();
  });

  function fetchElement(wrapper, index) {
    return wrapper.find('td').at(index).text();
  }

  it('column entries ok for 1 sample', () => {
    const wrapper = renderListView({...oneSample});
    const notick = fetchElement(wrapper, 0);
    expect(notick).toBe('');
    const sampleName = fetchElement(wrapper, 1);
    expect(sampleName).toBe('mysample');
    const location = fetchElement(wrapper, 2);
    expect(location).toBe('mycontainer');
    const sampleType = fetchElement(wrapper, 3);
    expect(sampleType).toBe('Zombie brain');
  });

  it('column entries ok for 1 group header', () => {
    const wrapper = renderListView({...oneSampleType});
    const tick = fetchElement(wrapper, 0);
    expect(tick).toBe('<CaretRight />');
    const sampleTypeName = fetchElement(wrapper, 1);
    expect(sampleTypeName).toBe('Zombie brain');
    const location = fetchElement(wrapper, 2);
    expect(location).toBe('');
    const sampleType = fetchElement(wrapper, 3);
    expect(sampleType).toBe('');
  });
});
