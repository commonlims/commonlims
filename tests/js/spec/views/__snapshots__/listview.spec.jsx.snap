// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`instantiate listview listView according to snapshot 1`] = `
ListView {
  "context": Object {},
  "props": Object {
    "allVisibleSelected": false,
    "canSelect": false,
    "columns": Array [
      Object {
        "Header": "Sample name",
        "accessor": [Function],
        "id": "name",
      },
      Object {
        "Header": "Container",
        "accessor": [Function],
        "id": "container",
      },
      Object {
        "Header": "Sample Type",
        "accessor": [Function],
        "id": "sample_type",
      },
    ],
    "dataById": Object {
      "1": Object {
        "entity": Object {
          "global_id": 1,
          "name": "Zombie brain",
        },
        "isGroupHeader": true,
      },
    },
    "errorMessage": "",
    "listActionBar": Object {},
    "loading": false,
    "selectedIds": Immutable.Set [],
    "toggleAll": [MockFunction],
    "toggleSingle": [MockFunction],
    "visibleIds": Array [
      1,
    ],
  },
  "refs": Object {},
  "state": Object {
    "expandedRows": Array [],
  },
  "updater": Updater {
    "_callbacks": Array [],
    "_renderer": ReactShallowRenderer {
      "_context": Object {},
      "_element": <ListView
        allVisibleSelected={false}
        canSelect={false}
        columns={
          Array [
            Object {
              "Header": "Sample name",
              "accessor": [Function],
              "id": "name",
            },
            Object {
              "Header": "Container",
              "accessor": [Function],
              "id": "container",
            },
            Object {
              "Header": "Sample Type",
              "accessor": [Function],
              "id": "sample_type",
            },
          ]
        }
        dataById={
          Object {
            "1": Object {
              "entity": Object {
                "global_id": 1,
                "name": "Zombie brain",
              },
              "isGroupHeader": true,
            },
          }
        }
        errorMessage=""
        listActionBar={Object {}}
        loading={false}
        selectedIds={Immutable.Set []}
        toggleAll={[MockFunction]}
        toggleSingle={[MockFunction]}
        visibleIds={
          Array [
            1,
          ]
        }
      />,
      "_forcedUpdate": false,
      "_instance": [Circular],
      "_newState": null,
      "_rendered": <Panel>
        Object {}
        <PanelBody
          direction="column"
          disablePadding={true}
          flex={false}
        >
          <Styles>
            <table>
              <thead>
                <tr>
                  <th>
                    <div />
                  </th>
                  <th>
                    <ColumnHeader>
                      Sample name
                    </ColumnHeader>
                  </th>
                  <th>
                    <ColumnHeader>
                      Container
                    </ColumnHeader>
                  </th>
                  <th>
                    <ColumnHeader>
                      Sample Type
                    </ColumnHeader>
                  </th>
                </tr>
              </thead>
              <tbody>
                <tr>
                  <td>
                    <CaretRight />
                  </td>
                  <td
                    style={
                      Object {
                        "fontStyle": "normal",
                      }
                    }
                  >
                    Zombie brain
                  </td>
                  <td
                    style={
                      Object {
                        "fontStyle": "normal",
                      }
                    }
                  />
                  <td
                    style={
                      Object {
                        "fontStyle": "normal",
                      }
                    }
                  />
                </tr>
              </tbody>
            </table>
          </Styles>
        </PanelBody>
      </Panel>,
      "_rendering": false,
      "_updater": [Circular],
    },
  },
}
`;
